benchmark/test/test_framework.py.in

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/env python3
+import subprocess
+import difflib
+import json
+import typing
+import re
+import pathlib
+import sys
+
+sourcepath = pathlib.Path("@CMAKE_CURRENT_SOURCE_DIR@")
+binpath = pathlib.Path("@PROJECT_BINARY_DIR@")
+generate = False
+if len(sys.argv) > 2 and sys.argv[2] == "--generate":
+    generate = True
+
+
+denumberify_paths = [
+    "time",
+    "bandwidth",
+    "flops",
+    "components",
+    "residual_norm",
+    "rhs_norm",
+    "max_relative_norm2",
+]
+empty_string_paths = ["error"]
+empty_array_paths = [
+    "recurrent_residuals",
+    "true_residuals",
+    "implicit_residuals",
+    "iteration_timestamps",
+]
+
+
+def sanitize_json_single(key, value, sanitize_all):
+    if key in empty_string_paths and isinstance(value, str):
+        return ""
+    if key in denumberify_paths and isinstance(value, float):
+        return 1.0
+    if key in denumberify_paths and isinstance(value, typing.Dict):
+        return sanitize_json(value, True)
+    if key in empty_array_paths and isinstance(value, typing.List):
+        return []
+    return sanitize_json(value, sanitize_all)
+
+
+def sanitize_json(parsed_input, sanitize_all=False):
+    if isinstance(parsed_input, typing.Dict):
+        return {
+            key: sanitize_json_single(key, value, sanitize_all)
+            for key, value in parsed_input.items()
+        }
+    elif isinstance(parsed_input, typing.List):
+        return [sanitize_json(e, sanitize_all) for e in parsed_input]
+    elif sanitize_all and isinstance(parsed_input, float):
+        return 1.0
+    else:
+        return parsed_input
+
+
+def sanitize_text(lines):
+    json_begins = [i for i, l in enumerate(lines) if l in ["[", "{"]]
+    json_ends = [i + 1 for i, l in enumerate(lines) if l in ["]", "}"]]
+    json_pairs = list(zip(json_begins, json_ends))
+    if len(json_pairs) == 0:
+        return lines
+    assert all(begin < end for begin, end in json_pairs)
+    nonjson_pairs = (
+        [(0, json_begins[0])]
+        + list(zip(json_ends[:-1], json_begins[1:]))
+        + [(json_ends[-1], len(lines))]
+    )
+    combined_pairs = sorted(
+        [(begin, end, False) for begin, end in nonjson_pairs]
+        + [(begin, end, True) for begin, end in json_pairs]
+    )
+    texts = [
+        ("\n".join(lines[begin:end]), do_sanitize)
+        for begin, end, do_sanitize in combined_pairs
+    ]
+    reconstructed = [
+        json.dumps(sanitize_json(json.loads(t)), indent=4) if do_sanitize else t
+        for t, do_sanitize in texts
+    ]
+    return "\n".join(reconstructed).split("\n")
```

Comment:
json.dumps will pretty-print the output and thus introduce newlines which mess up the text diff otherwise. I basically restore the invariant that each entry of the list represents a single line.

Reply:
That means that `reconstructed` contains "\n" ? Maybe add a quick comment, otherwise future us might remove that accidentally.